src/org/intellij/erlang/compilation/ErlangPrepareDependenciesCompileTask.java

```diff
@@ -128,161 +128,141 @@ private static ErlangModuleBuildOrderDescriptor getModuleBuildOrderInner(Module
     ErlangFacet erlangFacet = ErlangFacet.getFacet(module);
     List<String> globalParseTransforms = erlangFacet != null ? erlangFacet.getConfiguration().getParseTransforms() : ContainerUtil.<String>emptyList();
     buildOrder.myModuleName = module.getName();
-    buildOrder.myOrderedErlangModulePaths = getTopologicallySortedErlangModulePaths(ErlangModulesUtil.getErlangModules(module, false), globalParseTransforms);
-    buildOrder.myOrderedErlangTestModulePaths = getTopologicallySortedErlangModulePaths(ErlangModulesUtil.getErlangModules(module, true), ContainerUtil.<String>emptyList());
+    buildOrder.myOrderedErlangFilePaths = getTopologicallySortedErlangFilesPaths(ErlangModulesUtil.getErlangModules(module, false),
+                                                                                 ErlangModulesUtil.getErlangHeaders(module, false),
+                                                                                 globalParseTransforms);
+    buildOrder.myOrderedErlangTestFilePaths = getTopologicallySortedErlangFilesPaths(ErlangModulesUtil.getErlangModules(module, true),
+                                                                                     ErlangModulesUtil.getErlangHeaders(module, true),
+                                                                                     ContainerUtil.<String>emptyList());
     return buildOrder;
   }
 
-  private static List<ErlangFileDescriptor> getTopologicallySortedErlangModulePaths(Collection<ErlangFile> erlangModules,
-                                                                                    List<String> globalParseTransforms) throws CyclicDependencyFoundException {
-    return ErlangModulesSorter.sort(erlangModules, globalParseTransforms);
+  private static List<ErlangFileDescriptor> getTopologicallySortedErlangFilesPaths(@NotNull Collection<ErlangFile> erlangModules,
+                                                                                   @NotNull Collection<ErlangFile> erlangHeaders,
+                                                                                   @NotNull List<String> globalParseTransforms) throws CyclicDependencyFoundException {
```

Comment:
We can avoid having all of our PSI in memory. If you think about it, we only need at most two PSI files in memory at a time when computing dependencies. We can achieve this by operating on collections of VirtualFile instances and only requesting PSI from PsiManager when it's needed.